File: python/semanticabi/common/column/HexNormalize.py
from dataclasses import dataclass
from typing import Dict

from semanticabi.common.column.DatasetColumnTransform import DatasetColumnTransform


@dataclass
class HexNormalize(DatasetColumnTransform):
    """
    Lowercase hex strings.

    @author zuyezheng
    """

    # provide to rename
    source_col: str = None

    @staticmethod
    def normalize(hex: str) -> str:
        return hex.lower()

    def transform(self, blob: Dict[str, any], key: str) -> any:
        key = key if self.source_col is None else self.source_col

        if key in blob and blob[key] is not None:
            value = blob[key]
            if isinstance(value, list):
                return [HexNormalize.normalize(v) for v in value]
            else:
                return HexNormalize.normalize(value)
        else:
            return None

File: js/src/semanticabi/transform/column/HexNormalize.ts
import {DatasetColumnTransform} from "semanticabi/transform/column/DatasetColumnTransform";
import {JsonObject} from "common/CommonTypes";
import {Optional} from "common/Optional";

/**
 * Lowercase hex strings.
 *
 * @author zuyezheng
 */
export class HexNormalize implements DatasetColumnTransform {

    constructor(
        public readonly sourceCol: Optional<string> = Optional.none()
    ) { }

    static normalize(hex: string): string {
        if (hex == null) {
            return '';
        }

        return hex.toLowerCase();
    }

    transform(blob: JsonObject, key: string): string | string[] {
        return this.transformValue(blob[this.sourceCol.getOr(key)]);
    }

    transformValue(value: any): string | string[] {
        if (value == null) {
            return null;
        }

        if (Array.isArray(value)) {
            return value.map(HexNormalize.normalize);
        } else {
            return HexNormalize.normalize(value);
        }
    }

}